File: OC/显示层动画/WGOCAnimationPrimary/WGOCAnimationPrimary/旋转&缩放&平移/简单转盘/WGTurnPlateViewController.m
//
//  WGTurnPlateViewController.m
//  WGOCAnimationPrimary
//
//  Created by wanggang on 2019/5/21.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGTurnPlateViewController.h"

@interface WGTurnPlateViewController ()

@property (nonatomic, strong) UIImageView *imgView;
@property (nonatomic, assign) NSInteger index;
@property (nonatomic, assign) BOOL finished;

@end

@implementation WGTurnPlateViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"简单转盘";
    self.view.backgroundColor = UIColor.whiteColor;
    
    self.imgView = [[UIImageView alloc] initWithFrame:CGRectMake((SCREEN_WIDTH-300)/2, (SCREEN_HEIGHT-300)/2, 300, 300)];
    self.imgView.image = [UIImage imageNamed:@"turnPlate"];
    [self.view addSubview:self.imgView];
}

-(void)viewWillAppear:(BOOL)animated{
    
    [super viewWillAppear:animated];
    
    [UIView beginAnimations:nil context:nil];
    [UIView setAnimationDuration:0.3];
    [UIView setAnimationDelegate:self];
    self.imgView.transform = CGAffineTransformMakeRotation(M_PI_2);
    [UIView setAnimationDidStopSelector:@selector(animationStop)];
    [UIView commitAnimations];
}

- (void)viewDidDisappear:(BOOL)animated{
    [super viewDidDisappear:animated];
    
    self.finished = YES;
}

- (void)animationStop{
    
    if (!_finished) {
        
        [UIView beginAnimations:nil context:nil];
        [UIView setAnimationDuration:0.3];
        [UIView setAnimationDelegate:self];
        self.index += 1;
        CGFloat angle = M_PI_2*self.index;
        self.imgView.transform = CGAffineTransformMakeRotation(angle);
        [UIView setAnimationDidStopSelector:@selector(animationStop)];
        [UIView commitAnimations];
    }
}

@end

File: Swift/显示层初级动画/WGAnimationPrimary/WGAnimationPrimary/旋转&缩放&平移&回调/简单转盘/WGTurnPlateViewController.swift
//
//  WGTurnPlateViewController.swift
//  WGAnimationPrimary
//
//  Created by wanggang on 2019/4/22.
//  Copyright © 2019 bozhong. All rights reserved.
//

import UIKit

class WGTurnPlateViewController: UIViewController {

    var imgView: UIImageView?
    var index = 1.0
    //防止内存泄漏
    var isFinish = false
    
    override func viewDidLoad() {
        super.viewDidLoad()

        imgView = UIImageView()
        imgView?.bounds = CGRect(x: 0, y: 0, width: 300, height: 300)
        imgView?.center = self.view.center
        imgView?.image = UIImage(named: "turnPlate")
        imgView?.backgroundColor = UIColor.white
        self.view.backgroundColor = UIColor.white
        self.view.addSubview(imgView!)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        UIView.beginAnimations(nil, context: nil)
        UIView.setAnimationDelegate(self)
        UIView.setAnimationDuration(0.3)
        self.imgView?.transform = CGAffineTransform(rotationAngle: .pi/2)
        UIView.setAnimationDidStop(#selector(animationStop))
        UIView.commitAnimations()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        isFinish = true
    }
    
    @objc func animationStop() {
        
        if !isFinish {
            
            UIView.beginAnimations(nil, context: nil)
            UIView.setAnimationDelegate(self)
            UIView.setAnimationDuration(0.3)
            index += 1;
            let angle = (CGFloat)(Double.pi/2*index)
            self.imgView?.transform = CGAffineTransform(rotationAngle: angle)
            UIView.setAnimationDidStop(#selector(animationStop))
            UIView.commitAnimations()
        }
    }
}
